packages/calypso-url/src/wpcom-origin.ts

```diff
@@ -0,0 +1,58 @@
+/**
+ * These are the known places that Calypso is allowed to run.
+ */
+const ALLOWED_ORIGINS = [
+	'https://wordpress.com',
+	'https://horizon.wordpress.com',
+	'https://wpcalypso.wordpress.com',
+	'http://calypso.localhost:3000',
+	'https://calypso.localhost:3000',
+];
+
+/**
+ * Checks if the origin is allowed by checking against ALLOWED_ORIGINS.
+ *
+ * @param origin string to check
+ * @returns true if the origin is allowed
+ */
+function isAllowedOrigin( origin: string ) {
+	return ALLOWED_ORIGINS.includes( origin ) || isCalypsoLive( origin );
+}
+
+/**
+ * Checks if the origin is a Calypso Live site (e.g. https://busy-badger.calypso.live)
+ *
+ * @param origin string to check
+ * @returns true if the origin is a Calypso Live site
+ */
+function isCalypsoLive( origin: string ) {
+	return origin.match( /^https:\/\/[a-z0-9-]+\.calypso\.live$/ );
+}
+
+/**
+ * Gets the Calypso origin from the referrer or `origin` query arg
+ * and compares it against a list of allowed wpcom origins.
+ *
+ * @param path Optional path to append to the origin
+ * @returns The origin if it's allowed, otherwise the default origin (wordpress.com)
+ */
+export function getWpComOrigin( path = '' ) {
+	const defaultOrigin = 'https://wordpress.com';
+	const fromReferrer = document.referrer;
+	const fromQueryArg = new URLSearchParams( window.location.search ).get( 'origin' ) || '';
+	let origin = defaultOrigin;
+
+	if ( isAllowedOrigin( fromReferrer ) ) {
+		origin = fromReferrer;
+	} else if ( isAllowedOrigin( fromQueryArg ) ) {
+		origin = fromQueryArg;
+	}
```

Comment:
Do we ever set an "origin" query arg? Or is this just giving us future options?

nitpick - if we are checking both referrer and origin, should we bias the result to allow the origin to take precedence when valid? right now it reads as the opposite.

Reply:
Oh yeah good call. `origin` should take precedence with referrer as fallback in case it isn't set. [9b143a0](https://github.com/Automattic/wp-calypso/pull/74384/commits/9b143a0a3d1864cba5fe7725b0ef736f661b8e4a) in the removal branch adds the `origin` query arg to our outbound Editor URLs so they know what to come back to more readily. 

Then I'll have to open a Jetpack PR (underway) to maintain `origin=ORIGIN` query args across multiple screens, which will close the Post Editor --> Revisions --> Post Editor loop where env will be maintained. This will help any future testing effort where Calypso env consistency needs to be maintained.

With that in mind do you think `origin` is the best name? Seems vague, I just used it because I saw it was already being set in the iframe'd URL here.

Incidentally there are two ad-hoc places where 1. [origin is used](https://github.com/Automattic/wp-calypso/blob/23f817e32ccc5d8f7d555fb66b2bdc2dd057c87e/apps/editing-toolkit/editing-toolkit-plugin/wpcom-global-styles/index.php#L95-L109) and another where 2. [calypsoOrigin is used](https://github.com/Automattic/wp-calypso/search?q=calypsoOrigin).